x/lockup/keeper/synthetic_lock_test.go

```diff
@@ -0,0 +1,311 @@
+package keeper_test
+
+import (
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/osmosis-labs/osmosis/x/lockup/types"
+)
+
+func (suite *KeeperTestSuite) TestSyntheticLockupCreateGetDeleteAccumulation() {
+	suite.SetupTest()
+
+	// lock coins
+	addr1 := sdk.AccAddress([]byte("addr1---------------"))
+	coins := sdk.Coins{sdk.NewInt64Coin("stake", 10)}
+	suite.LockTokens(addr1, coins, time.Second)
+
+	// check locks
+	locks, err := suite.app.LockupKeeper.GetPeriodLocks(suite.ctx)
+	suite.Require().NoError(err)
+	suite.Require().Len(locks, 1)
+	suite.Require().Equal(locks[0].Coins, coins)
+
+	// check accumulation store is correctly updated
+	accum := suite.app.LockupKeeper.GetPeriodLocksAccumulation(suite.ctx, types.QueryCondition{
+		LockQueryType: types.ByDuration,
+		Denom:         "stake",
+		Duration:      time.Second,
+	})
+	suite.Require().Equal(accum.String(), "10")
+
+	// check queries for native denom before creating synthetic lockup
+	locks = suite.app.LockupKeeper.GetAccountLockedPastTimeDenom(suite.ctx, addr1, "stake", suite.ctx.BlockTime())
+	suite.Require().Len(locks, 1)
+	locks = suite.app.LockupKeeper.GetAccountLockedDurationNotUnlockingOnly(suite.ctx, addr1, "stake", time.Second)
+	suite.Require().Len(locks, 1)
+	locks = suite.app.LockupKeeper.GetAccountLockedLongerDurationDenom(suite.ctx, addr1, "stake", time.Second)
+	suite.Require().Len(locks, 1)
+	locks = suite.app.LockupKeeper.GetLocksPastTimeDenom(suite.ctx, "stake", suite.ctx.BlockTime())
+	suite.Require().Len(locks, 1)
+	locks = suite.app.LockupKeeper.GetLocksLongerThanDurationDenom(suite.ctx, "stake", time.Second)
+	suite.Require().Len(locks, 1)
+	amount := suite.app.LockupKeeper.GetLockedDenom(suite.ctx, "stake", time.Second)
+	suite.Require().Equal(amount.String(), "10")
+
+	// check queries for synthetic denom before creating synthetic lockup
+	locks = suite.app.LockupKeeper.GetAccountLockedPastTimeDenom(suite.ctx, addr1, "stakestakedtovalidator1", suite.ctx.BlockTime())
+	suite.Require().Len(locks, 0)
+	locks = suite.app.LockupKeeper.GetAccountLockedDurationNotUnlockingOnly(suite.ctx, addr1, "stakestakedtovalidator1", time.Second)
+	suite.Require().Len(locks, 0)
+	locks = suite.app.LockupKeeper.GetAccountLockedLongerDurationDenom(suite.ctx, addr1, "stakestakedtovalidator1", time.Second)
+	suite.Require().Len(locks, 0)
+	locks = suite.app.LockupKeeper.GetLocksPastTimeDenom(suite.ctx, "stakestakedtovalidator1", suite.ctx.BlockTime())
+	suite.Require().Len(locks, 0)
+	locks = suite.app.LockupKeeper.GetLocksLongerThanDurationDenom(suite.ctx, "stakestakedtovalidator1", time.Second)
+	suite.Require().Len(locks, 0)
+	amount = suite.app.LockupKeeper.GetLockedDenom(suite.ctx, "stakestakedtovalidator1", time.Second)
```

Comment:
Can we mitigate a lot of this code repetition in the tests via making other functions, with a struct for all the inputs, and a struct for all the expected answers?

Reply:
That way, code can be reduced but it won't be easy to read the code while reviewing tests.

Reply:
err, I don't think this style of tests is that easy to read, review, or give the guarantees we want. It ends up being very hard to adjust in the future / require rewriting. (Doesn't block this PR merge though)